src/accessTokenProvider.js

```diff
@@ -0,0 +1,111 @@
+import {
+  isResponseUnauthorized,
+} from './services/http';
+
+export class AccessTokenProvider {
+  constructor(fetch, config) {
+    this.fetch = fetch;
+
+    this.config = config;
+    this.refreshAccessTokenPromise = null;
+    this.tokens = {
+      refreshToken: null,
+      accessToken: null,
+    };
+
+    this.isAuthorized = this.isAuthorized.bind(this);
+    this.refresh = this.refresh.bind(this);
+    this.clear = this.clear.bind(this);
+
+    this.resolveAccessToken = this.resolveAccessToken.bind(this);
+    this.fetchToken = this.fetchToken.bind(this);
+    this.handleFetchResolved = this.handleFetchResolved.bind(this);
+    this.handleTokenResolved = this.handleTokenResolved.bind(this);
+    this.handleError = this.handleError.bind(this);
+  }
+
+  /**
+   * Refreshes current access token with provided refresh token
+   */
+  refresh() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      return Promise.resolve();
```

Comment:
If it rejects it will cause a chain reaction which will cause fetchIntercept chain to reject also, which in turn will reject an initial fetch request which is undesirable as we don't want to break the fetch spec which currently states that fetch will only reject on network errors and this is not a network error. 

But indeed it could be potentially problematic if you call `renew` without calling `authorize` first as it will silently ignore your `renew` requests. Two solutions here: silently fail with console error, or raise an exception which will be caught by fetchInterceptor which won't propagate it out of fetch call, but catch it, perhaps again log it, and then resolve as normal request would. I feel that the second solution only raises the handling to upper level without any flow control benefits as this can't be handled.